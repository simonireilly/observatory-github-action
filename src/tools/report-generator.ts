// Given a JSON report from the runner, it will translate it to markdown
import * as core from '@actions/core';

type JSONReport = {
  [key: string]: unknown;
};

interface Report {
  [key: string]: {
    score_modifier: string;
    pass: string;
    score_description: string;
  };
}

export function jsonReportToMarkdown(
  jsonReport: JSONReport | string,
  sanitizedHostName: string
): string {
  let result;

  if (typeof jsonReport === 'string') {
    const jsonStructure = jsonReport.slice(jsonReport.indexOf('{'));
    if (jsonStructure.length > 0) {
      result = JSON.parse(jsonStructure) as JSONReport;
    } else {
      core.setFailed('Result is empty');
      return '';
    }
  } else {
    result = jsonReport;
  }

  const resultRows: string[] = [];
  const report = result as Report;

  // Baseline score, points will be added or deducted
  let score = 100;

  // Get the keys
  for (const key in report) {
    const { score_modifier = '0', pass, score_description } = report[key];
    const success = Boolean(pass);

    score += parseInt(score_modifier);
    const icon = (showSuccessIcon: boolean): string =>
      showSuccessIcon ? ':green_circle:' : ':red_circle:';
    const message = `${icon(
      success
    )} | ${score_modifier} | ${score_description}`;
    resultRows.push(message);
  }

  return `
## Observatory Results [${sanitizedHostName}](https://${sanitizedHostName}): _${score} of 100_

See the full report: https://observatory.mozilla.org/analyze/${sanitizedHostName}

### Highlights

Passed | Score | Description
--- | --- | ---
${resultRows.join('\n')}
  `;
}
