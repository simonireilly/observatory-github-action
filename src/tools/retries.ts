export const runWithRetry = async (
  routineToRetry: () => Promise<number>,
  onExhaustedRetries: (e: unknown) => Promise<void>,
  maxRetries: number,
  baseSeconds = 1.6
) => {
  let attempts = 1;

  const retryRoutine = async (routine: () => Promise<number>): Promise<any> => {
    try {
      await routine();
    } catch (e) {
      if (attempts <= maxRetries) {
        await exponentialSetTimeout(attempts, baseSeconds);

        attempts += 1;

        return retryRoutine(routine);
      } else {
        return onExhaustedRetries(e);
      }
    }
  };

  return retryRoutine(routineToRetry);
};

export const exponentialSetTimeout = async (
  exponent: number,
  baseSeconds: number
) => {
  const exponentialMS = safeMillisecondExponent(exponent, baseSeconds);
  return new Promise((r) => setTimeout(r, 1000 + exponentialMS));
};

export const safeMillisecondExponent = (
  exponent: number,
  baseSeconds: number
): number => {
  return Math.floor(1000 * baseSeconds ** exponent);
};
