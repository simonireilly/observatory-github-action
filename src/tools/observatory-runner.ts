// Handlers spawning a process that runs the observatory CLI
import * as core from "@actions/core";
import * as exec from "@actions/exec";
import { runWithRetry } from "./retries";

export async function runObservatory(
  sanitizedHostName: string,
  maxRetries = 5
): Promise<{
  result: string;
  error?: string;
}> {
  // Setup scoped streams for piping into all attempts
  let result = "";
  let error = "";

  const options = {
    listeners: {},
  };

  options.listeners = {
    stdout: (data: Buffer) => {
      result += data.toString();
    },
    stderr: (data: Buffer) => {
      error += data.toString();
    },
  };

  await runWithRetry(
    () => runTool(sanitizedHostName, options),
    handleExhaustedRetries,
    maxRetries
  );

  core.info(result);
  core.error(error);

  return { result, error };
}

// Private helpers

const handleExhaustedRetries = async (e: unknown) => {
  if (e instanceof Error) {
    core.setFailed(e.message);
  } else {
    core.setFailed("An unknown error occurred");
  }
};

const runTool = async (
  sanitizedHostName: string,
  options: exec.ExecOptions
) => {
  return exec.exec(
    "node",
    [
      "node_modules",
      "observatory-cli",
      "index.js",
      sanitizedHostName,
      "--format=json",
      "--attempts=30",
    ],
    options
  );
};
