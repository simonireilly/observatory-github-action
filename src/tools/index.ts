// Handles configuration, and calling out to the runner
import { runObservatory } from './observatory-runner';
import * as core from '@actions/core';
import { jsonReportToMarkdown } from './report-generator';

const webHost = (): string => {
  return core.getInput('web_host') || 'github.com';
};

export async function run(): Promise<string> {
  let sanitizedHostName = webHost();

  try {
    sanitizedHostName = new URL(webHost()).host;
  } catch (e: unknown) {
    core.warning('This is not a valid URL, trying as given string');
    if (e instanceof Error) {
      core.error(e);
    } else {
      core.setFailed('An unknown error occurred');
    }
  }

  core.debug(`Running on website: ${sanitizedHostName}`);

  const { result, error } = await runObservatory(sanitizedHostName);

  if (error) core.debug(error);

  core.debug(result);

  const markdown = jsonReportToMarkdown(result, sanitizedHostName);

  core.setOutput('observatory-report', markdown);
  return markdown;
}
