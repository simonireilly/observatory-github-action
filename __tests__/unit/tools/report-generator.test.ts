import { jsonReportToMarkdown } from "../../../src/tools/report-generator";

import jsonReport from "../fixtures/json-observatory-report.json";
import * as fs from "fs";
import * as path from "path";

describe("jsonReportToMarkdown", () => {
  it("Creates a table for the results with json", () => {
    const markdown = jsonReportToMarkdown(jsonReport, "github.com");

    expect(markdown).toMatchSnapshot();
  });
  it("Creates a table for the results with text", () => {
    const jsonReport = fs.readFileSync(
      path.join(__dirname, "../fixtures/json-observatory-report.txt"),
      "utf8"
    );

    const markdown = jsonReportToMarkdown(jsonReport, "github.com");

    expect(markdown).toMatchSnapshot();
  });
});
