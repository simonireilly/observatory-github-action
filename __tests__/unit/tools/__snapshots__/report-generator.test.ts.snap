// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`processes a json file to a results object Creates a table for the results with json 1`] = `
Array [
  Array [
    ":red_circle:",
    "-25",
    "Content Security Policy (CSP) header not implemented",
  ],
  Array [
    ":green_circle:",
    "0",
    "Contribute.json isn't required on websites that don't belong to Mozilla",
  ],
  Array [
    ":green_circle:",
    "0",
    "No cookies detected",
  ],
  Array [
    ":green_circle:",
    "0",
    "Content is not visible via cross-origin resource sharing (CORS) files or headers",
  ],
  Array [
    ":green_circle:",
    "0",
    "HTTP Public Key Pinning (HPKP) header not implemented",
  ],
  Array [
    ":red_circle:",
    "-20",
    "Does not redirect to an HTTPS site",
  ],
  Array [
    ":green_circle:",
    "0",
    "Referrer-Policy header not implemented",
  ],
  Array [
    ":red_circle:",
    "-20",
    "HTTP Strict Transport Security (HSTS) header not implemented",
  ],
  Array [
    ":green_circle:",
    "0",
    "Subresource Integrity (SRI) is not needed since site contains no script tags",
  ],
  Array [
    ":red_circle:",
    "-5",
    "X-Content-Type-Options header not implemented",
  ],
  Array [
    ":red_circle:",
    "-20",
    "X-Frame-Options (XFO) header not implemented",
  ],
  Array [
    ":red_circle:",
    "-10",
    "X-XSS-Protection header not implemented",
  ],
]
`;

exports[`processes a json file to a results object processes a text file to a results object 1`] = `
Array [
  Array [
    ":red_circle:",
    "-25",
    "Content Security Policy (CSP) header not implemented",
  ],
  Array [
    ":green_circle:",
    "0",
    "Contribute.json isn't required on websites that don't belong to Mozilla",
  ],
  Array [
    ":green_circle:",
    "0",
    "No cookies detected",
  ],
  Array [
    ":green_circle:",
    "0",
    "Content is not visible via cross-origin resource sharing (CORS) files or headers",
  ],
  Array [
    ":green_circle:",
    "0",
    "HTTP Public Key Pinning (HPKP) header not implemented",
  ],
  Array [
    ":red_circle:",
    "-20",
    "Does not redirect to an HTTPS site",
  ],
  Array [
    ":green_circle:",
    "0",
    "Referrer-Policy header not implemented",
  ],
  Array [
    ":red_circle:",
    "-20",
    "HTTP Strict Transport Security (HSTS) header not implemented",
  ],
  Array [
    ":green_circle:",
    "0",
    "Subresource Integrity (SRI) is not needed since site contains no script tags",
  ],
  Array [
    ":red_circle:",
    "-5",
    "X-Content-Type-Options header not implemented",
  ],
  Array [
    ":red_circle:",
    "-20",
    "X-Frame-Options (XFO) header not implemented",
  ],
  Array [
    ":red_circle:",
    "-10",
    "X-XSS-Protection header not implemented",
  ],
]
`;
